docs/developer/decision-records/2024-10-09-vault-authentication-refactor/README.md

```diff
@@ -0,0 +1,223 @@
+# Hashicorp Vault authentication refactor
+
+## Decision
+
+Refactor the HashiCorp Vault extension and make it extensible with different methods of authentication.
+
+## Rationale
+
+The current implementation of the authentication in the HashiCorp Vault extension has no way to add new authentication methods and to switch between them.
+The full list of possible authentication methods can be found in the [HashiCorp Vault Documentation](https://developer.hashicorp.com/vault/docs/auth)
+Relevant examples are:
+
+* [Token auth](https://developer.hashicorp.com/vault/docs/auth/token)
+* [Kubernetes auth](https://developer.hashicorp.com/vault/docs/auth/kubernetes)
+* [AppRole auth](https://developer.hashicorp.com/vault/docs/auth/approle)
+It goes against the EDC extensibility model and needs to be remedied.
+Additionally, extracting the current implementation of the token authentication into its own extension will improve readability and maintainability of the code.
+
+## Approach
+The refactor will affect only the `vault-hashicorp` extension.
+
+To allow an extensible authentication module for HashiCorp Vault, the implementation will follow the registry pattern.
+
+### Vault Auth SPI
+
+To allow the proper organisation of the interfaces needed for the refactoring, a new module named `vault-auth-spi` is introduced in the `spi` directory.
+
+### Vault Authentication Service Interface
+
+To interact with a multitude of different authentication methods, an interface for the Vault Authentication Service is needed.
+This interface will contain two methods:
+
+* `login()`
+* `getVaultToken()`
+
+`login()` contains the authentication and fetches/generates the `client_token` from HashiCorp Vault.
+
+`getVaultToken()` returns the saved `client_token` when needed.
+
+```java
+public interface HashicorpVaultAuth {
+
+    void login();
+    
+    String getVaultToken();
+}
+```
+
+### Vault Authentication Service Implementation
+
+The Vault Authentication Service Implementations each implement a single HashiCorp Vault Authentication method.
+To achieve this, each extension contains an implementation of the newly introduced Vault Authentication Service interface.
+The goal of each extension is to authenticate with HashiCorp Vault and to provide a valid `client_token` to the `HashicorpVaultClient`.
+
+Interface Implementation:
+
+```java
+public record HashicorpVaultTokenAuth(@NotNull String vaultToken) implements HashicorpVaultAuth {
+
+    private String vaultToken;
+    
+    @Override
+    public void login() {
+        // no login is needed for token authentication
+    }
+
+    @Override
+    public String getVaultToken() {
```

Comment:
if the base method is called `vaultToken` there will be no need for this override